s2spy/preprocess.py

```diff
@@ -0,0 +1,217 @@
+
+
+def _check_input_data(data: Union[xr.DataArray, xr.Dataset]):
+    """Check the input data for compatiblity with the preprocessor.
+
+    Args:
+        data: Data to validate.
+
+    Raises:
+        ValueError: If the input data is of the wrong type.
+        ValueError: If the input data does not have a 'time' dimension.
+    """
+    if not any(isinstance(data, dtype) for dtype in (xr.DataArray, xr.Dataset)):
+        raise ValueError(
+            "Input data has to be an xarray-DataArray or xarray-Dataset, "
+            f"not {type(data)}"
+        )
+    if "time" not in data.dims:
+        raise ValueError(
+            "Analysis is done of the 'time' dimension, but the input data"
+            f" only has dims: {data.dims}"
+        )
+
+
+class Preprocessor:
+    """Preprocessor for s2s data."""
+    def __init__(
+        self,
+        rolling_window_size: Union[int, None],
+        rolling_min_periods: int = 1,
+        detrend: Union[str, None] = "linear",
+        remove_climatology: bool = True,
+    ):
+        """Preprocessor for s2s data. Can detrend as well as deseasonalize.
+
+        On calling `.fit(data)`, the preprocessor will:
+         - Calculate the rolling mean of the input data.
+         - Calculate and store the trend of the rolling mean.
+         - Calculate and store the climatology of the rolling mean.
+
+        When calling `.transform(data)`, the preprocessor will:
+         - Remove the (stored) trend from a copy of the data.
+         - Remove the climatology from this detrended data.
+         - Return the detrended and deseasonalized data.
+
+        Args:
+            rolling_window_size: The size of the rolling window that will be applied
+                before calculating the trend and climatology. Setting this to 1 will
+                effectively skip this step.
+            rolling_min_periods: The minimum number of periods within a rolling window.
+                If higher than 1 (the default), NaN values will be present at the start
+                and end of the preprocessed data.
+            detrend: Which method to use for detrending. Currently the only method
+                supported is "linear". If you want to skip detrending, set this to None.
+            remove_climatology (optional): If you want to calculate and remove the
+                climatology of the data. Defaults to True.
+        """
+        self._window_size = rolling_window_size
+        self._min_periods = rolling_min_periods
+        self._detrend = detrend
+        self._remove_climatology = remove_climatology
+
+        self._climatology: Union[xr.DataArray, xr.Dataset]
+        self._trend: dict
+        self._is_fit = False
+
+    def fit(self, data: Union[xr.DataArray, xr.Dataset]) -> None:
+        """Fit this Preprocessor to input data.
+
+        Args:
+            data
+        """
+        _check_input_data(data)
+        if self._window_size is not None:
```

Comment:
```suggestion
        if self._window_size not in [None, 1]:
```
Perhaps we should also just write a check to make sure that the window size is >=1.